### DISCUSSION TITLE
Registering Subclass Fields

### DISCUSSION THREAD

**Discussion by MrCroller** [ANU-CHEEKI-BREEKI/com.anupackages.debugconsole#10] [Q&A] [answered]
Hello, How would I register the player settings subclass fields to change them via console?

```csharp
[DebugCommandPrefix("player")]
class Player : MonoBehaviour
{
    Settings settings;

    [Serializable]
    class Settings
    {
        public float Speed;
        public float JumpForce;
    }
}
```
So that it could be changed like this
`player.setting Speed 200`

**Comment by ANU-CHEEKI-BREEKI** [comment, marked as answer]
Hello, thank you for your question.

Although the console does not currently support accessing members of nested fields, you can achieve the desired result by registering the specific fields directly.  
Note that since the `Settings` class is not a `MonoBehaviour`, you will need to register both the class and its instance.  
To do this, the class must be made public, and the command names should be assigned manually.

```csharp
[assembly: RegisterDebugCommandTypes(typeof(Player.Settings))]

class Player : MonoBehaviour
{
    [SerializeField] private Settings settings;

    [Serializable]
    public class Settings
    {
        [DebugCommand("player.speed")] public float Speed;
        [DebugCommand("player.jumpForce")] public float JumpForce;
    }

    private void Awake()
    {
        DebugConsole.InstanceTargets.Register(settings);
    }
}
```

Then, the command will look like this:  
`player.speed 200`

Alternatively, you can register the fields at runtime, which avoids making the `Settings` class public:

```csharp
class Player : MonoBehaviour
{
    [SerializeField] private Settings settings;

    [Serializable]
    [DebugCommandPrefix("player")]
    class Settings
    {
        [DebugCommand] public float Speed;
        [DebugCommand] public float JumpForce;
    }

    private void Awake()
    {
        DebugConsole.InstanceTargets.Register(settings);
        DebugConsole.Commands.RegisterCommand(new FieldCommand(typeof(Settings).GetField("Speed"), settings));
        DebugConsole.Commands.RegisterCommand(new FieldCommand(typeof(Settings).GetField("JumpForce"), settings));
    }
}
```

You can also register commands using lambdas, but this requires creating both getters and setters.  
Unfortunately, the console does not currently support command overloads, so the names must differ:

```csharp
class Player : MonoBehaviour
{
    [SerializeField] private Settings settings;

    [Serializable]
    class Settings
    {
        public float Speed;
        public float JumpForce;
    }

    private void Awake()
    {
        DebugConsole.Commands.RegisterCommand<float>("player.speed.set", "", speed => settings.Speed = speed);
        DebugConsole.Commands.RegisterCommand<float>("player.speed.get", "", () => settings.Speed);

        DebugConsole.Commands.RegisterCommand<float>("player.jump-force.set", "", jumpForce => settings.JumpForce = jumpForce);
        DebugConsole.Commands.RegisterCommand<float>("player.jump-force.get", "", () => settings.JumpForce);
    }
}
```

Then, the commands will look like this:  
`player.speed.set 200`

**Reply by MrCroller** [reply]
Thank you for your detailed answer.

**Reply by ANU-CHEEKI-BREEKI** [reply]
I would appreciate it if you could leave feedback on the Asset Store and/or leave a star on GitHub if you liked the console. Thanks

https://assetstore.unity.com/packages/tools/utilities/anu-runtime-console-commander-259584
